src/utils/__tests__/async-test.js

```diff
@@ -1,14 +1,53 @@
 /* @flow strict-local */
 import { sleep, tryUntilSuccessful } from '../async';
+import { Lolex } from '../../__tests__/aux/lolex';
 
 describe('sleep', () => {
+  const lolex: Lolex = new Lolex();
+
+  afterAll(() => {
+    lolex.dispose();
+  });
+
+  afterEach(() => {
+    // clear any unset timers
+    lolex.clearAllTimers();
+  });
+
   test('waits for a given time in milliseconds', async () => {
     const expectedMs = 1000;
-    const start = Date.now();
-    await sleep(expectedMs);
-    const durationMs = Date.now() - start;
-    expect(expectedMs).toBeLessThanOrEqual(durationMs);
-    expect(durationMs).toBeLessThan(10 * expectedMs);
+
+    const sleepAsync = async () => {
+      const start = Date.now();
+      await sleep(expectedMs);
+      const actualMs = Date.now() - start;
+
+      return actualMs;
+    };
+    const sleepPromise: Promise<number> = sleepAsync();
+
+    // `sleepPromise` should have set up its timer before returning control.
+    expect(lolex.getTimerCount()).toBe(1);
+
+    const FUDGE_FACTOR = 1.01; // see discussion of [α], below
+    lolex.advanceTimersByTime(expectedMs * FUDGE_FACTOR);
+
+    // If this stalls, Jest will report failure.
+    const actualMs = await sleepPromise;
+
+    // `sleep` should sleep for the specified number of milliseconds (and not,
+    // for example, that many seconds or microseconds).
+    expect(expectedMs).toBeLessThanOrEqual(actualMs);
+    expect(actualMs).toBeLessThan(10 * expectedMs); // [α]
+
+    // [α] In theory, we can't be sure of this test; the time between the
+    // specified timeout expiring and a timeslice actually becoming available
+    // may be arbitrarily long. (To simulate this with Lolex, one can set
```

Comment:
Is this true with Lolex? It seems like one of the great advantages of systematically mocking out the clock as well as timers, provided the mock implementation delivers this, is that we could expect the timers to behave ideally.

Which effectively means that for timer-intensive code like this, we'd have written much more of a true *unit test* than was ever possible with real clocks.

Reply:
> Is this true with Lolex?

It's not – hence the trailing parenthetical. That could stand to be made more clear, though.

> It seems like one of the great advantages of systematically mocking out the clock as well as timers, provided the mock implementation delivers this, is that we could expect the timers to behave ideally.

Alas, "ideally" contrasts with "in reality", making this simultaneously one of the great *disadvantages* of mocking out the clock. We don't have any simple way to ensure that our code behaves properly under real-world conditions like load-induced greater-than-epsilon timeslice delays.

> Which effectively means that for timer-intensive code like this, we'd have written much more of a true *unit test* than was ever possible with real clocks.

That's true, but it's not an unalloyed good. :disappointed: An integration test may well be more informative than a unit test.